### PR TITLE
[WIP] Refactor Web UI code

### PR DESCRIPTION
**Pull request by tieujason330**
- created higher-order components (hoc) from common code throughout the various pages
   - examples: 
      - `withErrors` is an stateless hoc that will show errors instead of the page if any errors exist
      - `withFetchData` is a stateful hoc that manages when a component needs to send out a request
- most page views are turned into stateless components (no state, only renders)
- focused on shallow unit tests so we can test components individually
- will add integration tests later on

### CODE REVIEW

File: webui/common/src/components/FileView/FileView.tsx [hunk 1]
```diff
@@ -90,7 +81,7 @@ export class FileView extends React.PureComponent<IFileViewProps> {
             </ButtonGroup>
           </FormGroup>
           <FormGroup className="col-2">
-            <Button tag={Link} to={`${queryStringPrefix}?path=${path}${queryStringSuffix}`} color="secondary">Go</Button>
+            <Button tag={Link} to={`${queryStringPrefix}?path=${encodeURIComponent(path || '')}${queryStringSuffix}`} color="secondary">Go</Button>
```
Comment (calvinjia) [inline webui/common/src/components/FileView/FileView.tsx]: Was this previously a bug?

Comment (tieujason330) [reply webui/common/src/components/FileView/FileView.tsx]: the path is always decoded now and we encode it when it's used to navigate, so I added that to be consistent

File: webui/common/src/utilities/misc/createAlertErrors.tsx [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import {IAlertErrors} from "../../constants";
+
+export function createAlertErrors(errorCondition: boolean, errorList?: string[]): IAlertErrors {
+    const errors: IAlertErrors = {
+        hasErrors: errorCondition,
+        general: errorCondition,
+        specific: []
+    };
+
+    errorList && errorList.forEach(err => err && (errors.hasErrors = true) && errors.specific.push(err));
```
Comment (calvinjia) [inline webui/common/src/utilities/misc/createAlertErrors.tsx]: Is this the recommended programming style (instead of checking for undef/true)?

File: webui/common/src/utilities/misc/getDisplayName.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+import * as React from "react";
```
Comment (calvinjia) [inline webui/common/src/utilities/misc/getDisplayName.tsx]: Do we use `*` imports?

File: webui/common/src/utilities/misc/parseQueryString.tsx [hunk 4]
```diff
@@ -13,7 +13,7 @@ export interface IParsedQueryString {
   [key: string]: string;
 }
 
-export const parseQuerystring = (searchString: string): IParsedQueryString => {
+export const parseQueryString = (searchString: string): IParsedQueryString => {
```
Comment (calvinjia) [inline webui/common/src/utilities/misc/parseQueryString.tsx]: searchString -> queryString?

File: webui/common/src/utilities/misc/renderFileNameLink.tsx [hunk 5]
```diff
@@ -1,3 +1,14 @@
...
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
 import {Link} from 'react-router-dom';
 import React from 'react';
```
Comment (calvinjia) [inline webui/common/src/utilities/misc/renderFileNameLink.tsx]: Is there an import order we should follow? Would be nice to get all these style things in a linter, especially since we are committing so much code (should be done in a separate PR)

File: webui/common/src/utilities/saga/getSagaRequest.tsx [hunk 6]
```diff
@@ -11,6 +11,7 @@
 
 import {call, put} from 'redux-saga/effects';
 import {ActionType} from 'typesafe-actions';
+import {AxiosResponse} from 'axios';
```
Comment (calvinjia) [inline webui/common/src/utilities/saga/getSagaRequest.tsx]: Is this necessary?

File: webui/master/src/constants/data/headerNavigation.tsx [hunk 7]
```diff
@@ -10,26 +10,27 @@
...
-  url: '/overview'
+  url: routePaths.overview
 }, {
   innerText: 'Browse',
-  url: '/browse'
+  url: routePaths.browse
 }, {
   innerText: 'In-Alluxio Data',
-  url: '/data'
+  url: routePaths.data
```
Comment (calvinjia) [inline webui/master/src/constants/data/headerNavigation.tsx]: is `data` a reserved name?

File: webui/master/src/constants/data/routePaths.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+export const routePaths = {
+    root: '/',
+    overview: '/overview',
+    browse: '/browse',
+    config: '/config',
+    data: '/data',
```
Comment (calvinjia) [inline webui/master/src/constants/data/routePaths.tsx]: Unrelated to this PR - we should probably change this path to be more descriptive.

File: webui/master/src/containers/App/App.tsx [hunk 9]
```diff
@@ -9,100 +9,77 @@
...
+import {compose, Dispatch} from 'redux';
 
-import {Footer, Header, LoadingMessage} from '@alluxio/common-ui/src/components';
+import {
+  Footer,
+  withErrors,
+  withLoadingMessage,
+  Header,
+  withFetchData
+} from '@alluxio/common-ui/src/components';
```
Comment (calvinjia) [inline webui/master/src/containers/App/App.tsx]: one line?

File: webui/master/src/containers/pages/Browse/Browse.test.tsx [hunk 10]
```diff
@@ -53,39 +60,57 @@ describe('Browse', () => {
...
+  //   });
+  //
+  //   it('Calls fetchRequest', () => {
+  //     sinon.assert.called(props.fetchRequest as SinonSpy);
+  //   });
+  //
+  //   it('Matches snapshot', () => {
+  //     expect(reactWrapper).toMatchSnapshot();
+  //   });
+  // });
```
Comment (calvinjia) [inline webui/master/src/containers/pages/Browse/Browse.test.tsx]: Should we delete this or is this expected to work later? (applies in general to commented out code)

File: webui/common/src/utilities/misc/createAlertErrors.tsx (hunk 2 above)
Comment (tieujason330) [reply webui/common/src/utilities/misc/createAlertErrors.tsx]: this is shorthand but i can make it more clear

File: webui/common/src/utilities/misc/getDisplayName.tsx (hunk 3 above)
Comment (tieujason330) [reply webui/common/src/utilities/misc/getDisplayName.tsx]: updated

File: webui/master/src/constants/data/routePaths.tsx (hunk 8 above)
Comment (tieujason330) [reply webui/master/src/constants/data/routePaths.tsx]: noted

File: webui/master/src/containers/pages/Browse/Browse.test.tsx (hunk 10 above)
Comment (tieujason330) [reply webui/master/src/containers/pages/Browse/Browse.test.tsx]: cleaned up these commented blocks

File: webui/common/src/components/Logs/Logs.test.tsx [hunk 11]
```diff
@@ -40,7 +40,7 @@ describe('Logs', () => {
       loading: false,
       class: '',
       errors: createAlertErrors(false),
-      upateRequestParameter: sinon.spy()
+      update: sinon.spy()
```
Comment (calvinjia) [inline webui/common/src/components/Logs/Logs.test.tsx]: I meant to change it to `updateRequestParameter`, because `update` was misspelled

Comment (tieujason330) [reply webui/common/src/components/Logs/Logs.test.tsx]: ohh woops - updated

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.
